File: jbake-core/src/test/java/org/jbake/render/DocumentsRendererTest.java
package org.jbake.render;

import org.jbake.app.ContentStore;
import org.jbake.app.DocumentList;
import org.jbake.app.Renderer;
import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.model.DocumentModel;
import org.jbake.model.DocumentTypes;
import org.jbake.model.ModelAttributes;
import org.jbake.template.RenderingException;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.jupiter.api.Assertions;
import org.junit.rules.ExpectedException;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.MockitoAnnotations;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class DocumentsRendererTest {

    public DocumentsRenderer documentsRenderer;
    private ContentStore db;
    private Renderer renderer;
    private JBakeConfiguration configuration;
    private DocumentList<DocumentModel> emptyTemplateModelList;

    @Captor
    private ArgumentCaptor<DocumentModel> argument;

    @Before
    public void setUp() {

        MockitoAnnotations.initMocks(this);

        documentsRenderer = new DocumentsRenderer();

        db = mock(ContentStore.class);
        renderer = mock(Renderer.class);
        configuration = mock(JBakeConfiguration.class);
        emptyTemplateModelList = new DocumentList<>();
    }

    @Test
    public void shouldReturnZeroIfNothingHasRendered() throws Exception {

        when(db.getUnrenderedContent()).thenReturn(emptyTemplateModelList);

        int renderResponse = documentsRenderer.render(renderer, db, configuration);

        assertThat(renderResponse).isEqualTo(0);
    }

    @Test
    public void shouldReturnCountOfProcessedDocuments() throws Exception {

        // given:
        DocumentTypes.addDocumentType("customType");

        DocumentList<DocumentModel> templateModelList = new DocumentList<>();
        templateModelList.add(emptyDocument());
        templateModelList.add(emptyDocument());

        // return given DocumentList for DocumentType 'custom type'
        when(db.getUnrenderedContent()).thenReturn(templateModelList);
        when(db.getAllContent(any())).thenReturn(templateModelList);

        // when:
        int renderResponse = documentsRenderer.render(renderer, db, configuration);

        // then:
        assertThat(renderResponse).isEqualTo(2);
    }

    @Test
    public void shouldThrowAnExceptionWithCollectedErrorMessages() {
        String fakeExceptionMessage = "fake exception";

        // expect
        Assertions.assertThrows(
            RenderingException.class, () -> {

                // given
                DocumentTypes.addDocumentType("customType");

                DocumentList<DocumentModel> templateModelList = new DocumentList<>();
                DocumentModel document = emptyDocument();
                DocumentModel document2 = emptyDocument();
                templateModelList.add(document);
                templateModelList.add(document2);

                // throw an exception for every call of renderer's render method
                doThrow(new Exception(fakeExceptionMessage)).when(renderer).render(any(DocumentModel.class));
                when(db.getUnrenderedContent()).thenReturn(templateModelList);

                // when
                int renderResponse = documentsRenderer.render(renderer, db, configuration);

                // then
                assertThat(renderResponse).isEqualTo(2);
            },
            fakeExceptionMessage + "\n" + fakeExceptionMessage
        );
    }

    @Test
    public void shouldContainPostNavigation() throws Exception {
        // given
        DocumentTypes.addDocumentType("customType");

        String firstTitle = "First Document";
        DocumentModel firstDoc = simpleDocument(firstTitle, ModelAttributes.Status.PUBLISHED, "page");
        String secondTitle = "Second Document";
        DocumentModel secondDoc = simpleDocument(secondTitle, ModelAttributes.Status.PUBLISHED, "post");
        String thirdTitle = "Third Document";
        DocumentModel thirdDoc = simpleDocument(thirdTitle, ModelAttributes.Status.PUBLISHED, "page");
        String fourthTitle = "Fourth Document (draft)";
        DocumentModel fourthDoc = simpleDocument(fourthTitle, ModelAttributes.Status.DRAFT, "post");
        String fifthTitle = "Fifth Document";
        DocumentModel fifthDoc = simpleDocument(fifthTitle, ModelAttributes.Status.PUBLISHED, "page");
        String sixthTitle = "Sixth Document";
        DocumentModel sixthDoc = simpleDocument(sixthTitle, ModelAttributes.Status.PUBLISHED, "post");
        String seventhTitle = "Seventh Document";
        DocumentModel seventhDoc = simpleDocument(seventhTitle, ModelAttributes.Status.PUBLISHED, "post");

        DocumentList allDocs = new DocumentList();
        allDocs.add(seventhDoc);
        allDocs.add(sixthDoc);
        allDocs.add(fifthDoc);
        allDocs.add(fourthDoc);
        allDocs.add(thirdDoc);
        allDocs.add(secondDoc);
        allDocs.add(firstDoc);

        DocumentList pageDocs = new DocumentList();
        pageDocs.add(fifthDoc);
        pageDocs.add(thirdDoc);
        pageDocs.add(firstDoc);

        DocumentList postDocs = new DocumentList();
        postDocs.add(seventhDoc);
        postDocs.add(sixthDoc);
        postDocs.add(fourthDoc);
        postDocs.add(secondDoc);

        when(db.getUnrenderedContent()).thenReturn(allDocs);
        when(db.getAllContent("page")).thenReturn(pageDocs);
        when(db.getAllContent("post")).thenReturn(postDocs);

        // when
        int renderResponse = documentsRenderer.render(renderer, db, configuration);

        // then
        verify(renderer, times(7)).render(argument.capture());
        final Map<String, Map<String, Object>> renderedDocs = asTitleToDocMap(argument.getAllValues());

        // page checks
        assertDocumentNavigation(renderedDocs.get(fifthTitle), thirdTitle, null);
        assertDocumentNavigation(renderedDocs.get(thirdTitle), firstTitle, fifthTitle);
        assertDocumentNavigation(renderedDocs.get(firstTitle), null, thirdTitle);

        // post checks
        assertDocumentNavigation(renderedDocs.get(seventhTitle), sixthTitle, null);
        assertDocumentNavigation(renderedDocs.get(sixthTitle), secondTitle, seventhTitle);
        assertDocumentNavigation(renderedDocs.get(fourthTitle), secondTitle, sixthTitle);
        assertDocumentNavigation(renderedDocs.get(secondTitle), null, sixthTitle);

        assertThat(renderResponse).isEqualTo(7);
    }

    private void assertDocumentNavigation(
        final Map<String, Object> renderedDoc,
        final String prevDocumentTitle, String nextDocumentTitle) {
        assertThat(renderedDoc).flatExtracting(
            "previousContent." + ModelAttributes.TITLE,
            "nextContent." + ModelAttributes.TITLE)
            .containsExactly(prevDocumentTitle, nextDocumentTitle);
    }

    private Map<String, Map<String, Object>> asTitleToDocMap(List<DocumentModel> values) {
        return values.stream()
            .collect(Collectors.toMap(doc -> doc.get(ModelAttributes.TITLE).toString(), doc -> doc));
    }

    private DocumentModel emptyDocument() {
        return new DocumentModel();
    }

    private DocumentModel simpleDocument(String title, String status, String docType) {
        DocumentModel simpleDoc = new DocumentModel();
        String uri = title.replace(" ", "_");
        simpleDoc.setNoExtensionUri(uri);
        simpleDoc.setUri(uri);
        simpleDoc.setType(docType);
        simpleDoc.setTitle(title);
        simpleDoc.setStatus(status);
        return simpleDoc;
    }

}

File: jbake-core/src/test/java/org/jbake/render/DocumentsRendererTest.kt
package org.jbake.render

import org.assertj.core.api.Assertions
import org.jbake.app.ContentStore
import org.jbake.app.DocumentList
import org.jbake.app.Renderer
import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.model.DocumentModel
import org.jbake.model.DocumentTypes.addDocumentType
import org.jbake.model.ModelAttributes
import org.jbake.template.RenderingException
import org.junit.Before
import org.junit.Test
import org.junit.jupiter.api.assertThrows
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.*

// Helper functions for Kotlin null-safety with Mockito
fun <T> capture(captor: ArgumentCaptor<T>): T = captor.capture() ?: null as T
@Suppress("UNCHECKED_CAST")
fun <T> any(type: Class<T>): T {
    org.mockito.ArgumentMatchers.any(type)
    return null as T
}
@Suppress("UNCHECKED_CAST")
fun <T> anyNullable(type: Class<T>): T {
    org.mockito.ArgumentMatchers.nullable(type)
    return null as T
}

class DocumentsRendererTest {
    lateinit var documentsRenderer: DocumentsRenderer
    private lateinit var db: ContentStore
    private lateinit var renderer: Renderer
    private lateinit var configuration: JBakeConfiguration
    private lateinit var emptyTemplateModelList: DocumentList<DocumentModel>
    private lateinit var argument: ArgumentCaptor<DocumentModel>

    @Before
    fun setUp() {
        documentsRenderer = DocumentsRenderer()

        db = mock(ContentStore::class.java)
        renderer = mock(Renderer::class.java)
        configuration = mock(JBakeConfiguration::class.java)
        emptyTemplateModelList = DocumentList()
        argument = ArgumentCaptor.forClass(DocumentModel::class.java)
    }

    @Test fun shouldReturnZeroIfNothingHasRendered() {
        `when`(db.unrenderedContent).thenReturn(emptyTemplateModelList)

        val renderResponse = documentsRenderer.render(renderer, db, configuration)

        Assertions.assertThat(renderResponse).isEqualTo(0)
    }

    @Test fun shouldReturnCountOfProcessedDocuments() {
        // given:
        addDocumentType("customType")

        val templateModelList: DocumentList<DocumentModel> = DocumentList()
        templateModelList.add(emptyDocument())
        templateModelList.add(emptyDocument())

        // return given DocumentList for DocumentType 'custom type'
        `when`(db.unrenderedContent).thenReturn(templateModelList)
        `when`(db.getAllContent(any(String::class.java))).thenReturn(templateModelList)

        // when:
        val renderResponse = documentsRenderer.render(renderer, db, configuration)

        // then:
        Assertions.assertThat(renderResponse).isEqualTo(2)
    }

    @Test fun shouldThrowAnExceptionWithCollectedErrorMessages() {
        val fakeExceptionMessage = "fake exception"

        // given
        addDocumentType("customType")

        val templateModelList: DocumentList<DocumentModel> = DocumentList()
        val document = emptyDocument()
        val document2 = emptyDocument()
        templateModelList.add(document)
        templateModelList.add(document2)

        // throw an exception for every call of renderer's render method
        doThrow(RuntimeException(fakeExceptionMessage))
            .`when`(renderer).render(anyNullable(DocumentModel::class.java))

        `when`(db.unrenderedContent).thenReturn(templateModelList)
        `when`(db.getAllContent(any(String::class.java))).thenReturn(templateModelList)

        // expect
        val exception = assertThrows<RenderingException> { documentsRenderer.render(renderer, db, configuration) }
        Assertions.assertThat(exception.message).contains(fakeExceptionMessage)
    }

    @Test fun shouldContainPostNavigation() {
        // given
        addDocumentType("customType")

        val firstTitle = "First Document"
        val firstDoc = simpleDocument(firstTitle, ModelAttributes.Status.PUBLISHED, "page")
        val secondTitle = "Second Document"
        val secondDoc = simpleDocument(secondTitle, ModelAttributes.Status.PUBLISHED, "post")
        val thirdTitle = "Third Document"
        val thirdDoc = simpleDocument(thirdTitle, ModelAttributes.Status.PUBLISHED, "page")
        val fourthTitle = "Fourth Document (draft)"
        val fourthDoc = simpleDocument(fourthTitle, ModelAttributes.Status.DRAFT, "post")
        val fifthTitle = "Fifth Document"
        val fifthDoc = simpleDocument(fifthTitle, ModelAttributes.Status.PUBLISHED, "page")
        val sixthTitle = "Sixth Document"
        val sixthDoc = simpleDocument(sixthTitle, ModelAttributes.Status.PUBLISHED, "post")
        val seventhTitle = "Seventh Document"
        val seventhDoc = simpleDocument(seventhTitle, ModelAttributes.Status.PUBLISHED, "post")

        val allDocs = DocumentList<DocumentModel>()
        allDocs.add(seventhDoc)
        allDocs.add(sixthDoc)
        allDocs.add(fifthDoc)
        allDocs.add(fourthDoc)
        allDocs.add(thirdDoc)
        allDocs.add(secondDoc)
        allDocs.add(firstDoc)

        val pageDocs = DocumentList<DocumentModel>()
        pageDocs.add(fifthDoc)
        pageDocs.add(thirdDoc)
        pageDocs.add(firstDoc)

        val postDocs = DocumentList<DocumentModel>()
        postDocs.add(seventhDoc)
        postDocs.add(sixthDoc)
        postDocs.add(fourthDoc)
        postDocs.add(secondDoc)

        `when`(db.unrenderedContent).thenReturn(allDocs)
        `when`(db.getAllContent("page")).thenReturn(pageDocs)
        `when`(db.getAllContent("post")).thenReturn(postDocs)

        // when
        val renderResponse = documentsRenderer.render(renderer, db, configuration)

        // then
        verify(renderer, times(7)).render(capture(argument))
        val renderedDocs = asTitleToDocMap(argument.getAllValues())

        // page checks
        assertDocumentNavigation(renderedDocs[fifthTitle], thirdTitle, null)
        assertDocumentNavigation(renderedDocs[thirdTitle], firstTitle, fifthTitle)
        assertDocumentNavigation(renderedDocs[firstTitle], null, thirdTitle)

        // post checks
        assertDocumentNavigation(renderedDocs[seventhTitle], sixthTitle, null)
        assertDocumentNavigation(renderedDocs[sixthTitle], secondTitle, seventhTitle)
        assertDocumentNavigation(renderedDocs[fourthTitle], secondTitle, sixthTitle)
        assertDocumentNavigation(renderedDocs[secondTitle], null, sixthTitle)

        Assertions.assertThat(renderResponse).isEqualTo(7)
    }

    private fun assertDocumentNavigation(
        renderedDoc: MutableMap<String,  Any>?,
        prevDocumentTitle: String?, nextDocumentTitle: String?
    ) {
        Assertions.assertThat(renderedDoc).flatExtracting(
            "previousContent." + ModelAttributes.TITLE,
            "nextContent." + ModelAttributes.TITLE
        )
            .containsExactly(prevDocumentTitle, nextDocumentTitle)
    }

    private fun asTitleToDocMap(values: MutableList<DocumentModel>): MutableMap<String, MutableMap<String,  Any>> {
        return values.associateBy { it.get(ModelAttributes.TITLE).toString() }.toMutableMap()
    }

    private fun emptyDocument(): DocumentModel {
        return DocumentModel()
    }

    private fun simpleDocument(title: String, status: String?, docType: String): DocumentModel {
        val simpleDoc = DocumentModel()
        val uri = title.replace(" ", "_")
        simpleDoc.noExtensionUri = uri
        simpleDoc.uri = uri
        simpleDoc.type = docType
        simpleDoc.title = title
        simpleDoc.status = status
        return simpleDoc
    }
}
